### PR TITLE
Fix compiler warnings and hlint suggestions.

### PR DESCRIPTION
**Pull request by philderbeast**
I tested with ghc-8.10.7 and ghc-9.2.4 with stack and cabal, fixed the compiler warnings (except the missing type signatures in the tests that I ignored), added hlint testing and followed hlint's suggestions.

### CODE REVIEW

**Comment by saurabhnanda** [comment]
Thank you for the PR @philderbeast 

Are you using odd-jobs in production? Are `hlint` suggestions something that you _really_ care about?

The reason I'm asking is, if that I merge any other PR before yours, your PR will become outdated. And the diff of your PR is very large and will require significant rework.

Any thoughts @jappeace @ivb-supercede ?

File: ghc-9.2.4.stack.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+resolver: nightly-2022-10-11
+packages:
+  - .
```
Comment (jappeace) [inline ghc-9.2.4.stack.yaml:3]: maybe also add 9.2 support in the ci, since you went trough the effort of figuring out the stack.yaml file.

**Review by jappeace — APPROVED** [review]
Seems like a good idea, asside from silently dropping ghc 8.8 support in CI, I'd no add hlint to my CI pipeline as something to fail upon, but that's a personal preference thing. (Internally we've it configured as just a warning).

**Comment by philderbeast** [comment]
> Are `hlint` suggestions something that you _really_ care about?

I really like hlint, [contributed counts](https://github.com/ndmitchell/hlint/pull/1380) to the `hlint --default` command, use it in my projects have introduced it to other projects, [liquid-fixpoint](https://github.com/ucsd-progsys/liquid-fixpoint) being the biggest one.

File: ghc-9.2.4.stack.yaml (hunk 1 above)
Comment (philderbeast) [reply ghc-9.2.4.stack.yaml]: I bumped `nightly` to the `LTS` resolver for `ghc-9.2.5`.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -14,8 +14,8 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        stack: ["2.3.3"]
-        ghc: ["8.8.3"]
```
[the remark this answers is not in the document]
Comment (philderbeast) [reply .github/workflows/ci.yml:18]: I gave a modest version bump. Let me know if you'd like me to add a `ghc-8.8` version, say `ghc-8.8.4` the latest in that series.

**Comment by philderbeast** [comment]
Here's the counts running `hlint-3.5` on the master branch:

```yaml
# Warnings currently triggered by your code
- ignore: {name: "Eta reduce"} # 6 hints
- ignore: {name: "Move brackets to avoid $"} # 19 hints
- ignore: {name: "Redundant $"} # 21 hints
- ignore: {name: "Redundant <$>"} # 1 hint
- ignore: {name: "Redundant bracket"} # 121 hints
- ignore: {name: "Redundant fromIntegral"} # 1 hint
- ignore: {name: "Redundant id"} # 1 hint
- ignore: {name: "Redundant if"} # 1 hint
- ignore: {name: "Redundant lambda"} # 1 hint
- ignore: {name: "Replace case with fromMaybe"} # 2 hints
- ignore: {name: "Unused LANGUAGE pragma"} # 4 hints
- ignore: {name: "Use <$>"} # 6 hints
- ignore: {name: "Use <&>"} # 1 hint
- ignore: {name: "Use asks"} # 12 hints
- ignore: {name: "Use const"} # 2 hints
- ignore: {name: "Use fewer imports"} # 2 hints
- ignore: {name: "Use if"} # 1 hint
- ignore: {name: "Use join"} # 1 hint
- ignore: {name: "Use lambda-case"} # 5 hints
- ignore: {name: "Use list literal pattern"} # 2 hints
- ignore: {name: "Use newtype instead of data"} # 1 hint
- ignore: {name: "Use null"} # 3 hints
- ignore: {name: "Use optional"} # 1 hint
- ignore: {name: "Use void"} # 1 hint
``

**Comment by ivb-supercede** [comment]
I like fixing lints and warnings, they always bug me when building the project from source. I'm happy to merge this if nobody objects soon (but nobody has objected so far, so I won't wait very long.)

File: .github/workflows/ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jappeace) [reply .github/workflows/ci.yml:18]: yes please add it back, 8.8 to 8.10 is a major release, the ghc versioning is weird like that

**Comment by jappeace** [comment]
please merge yes.

File: .github/workflows/ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivb-supercede) [reply .github/workflows/ci.yml:18]: I'll do this on the main branch.

**Pull request merged**